tests/test_rand.py

```diff
@@ -8,3 +15,189 @@ def test_random():
+            assert rngs == child_rngs
+
+    else:
+        # child
+        try:
+            rngs = {_rand.rand64bits() for _ in range(10000)}
+            q.put(rngs)
+            q.get()
+        finally:
+            # Kill the process so it doesn't continue running the rest of the
+            # test suite in a separate process. Note we can't use sys.exit()
+            # as it raises an exception that pytest will detect as an error.
+            os._exit(0)
+
+
+def test_patching_random_seed():
+    import random
+
+    seed = _rand._getstate()
+    random.seed()
+    assert _rand._getstate() != seed
+
+
+def test_multiprocess():
+    q = mp.Queue()
+
+    def target(q):
+        q.put([_rand.rand64bits() for _ in range(100)])
+
+    ps = [mp.Process(target=target, args=(q,)) for _ in range(10)]
+    for p in ps:
+        p.start()
+
+    for p in ps:
+        p.join()
+
+    ids_list = [_rand.rand64bits() for _ in range(10000)]
+    ids = set(ids_list)
+    assert len(ids_list) == len(ids), "Collisions found in ids"
+
+    while not q.empty():
+        child_ids_list = q.get()
+        child_ids = set(child_ids_list)
+
+        assert len(child_ids_list) == len(child_ids), "Collisions found in subprocess ids"
+
+        assert ids & child_ids == set()
+        ids = ids | child_ids  # accumulate the ids
+
+
+def test_threadsafe():
+    # Check that the PRNG is thread-safe.
+    # This obviously won't guarantee thread safety, but it's something
+    # at least.
+    # To provide some validation of this method I wrote a slow, unsafe RNG:
+    #
+    # state = 4101842887655102017
+    #
+    # def bad_random():
+    #     global state
+    #     state ^= state >> 21
+    #     state ^= state << 35
+    #     state ^= state >> 4
+    #     return state * 2685821657736338717
+    #
+    # which consistently fails this test.
+
+    q = Queue()
+
+    def _target():
+        # Generate a bunch of numbers to try to maximize the chance that
+        # two threads will be calling rand64bits at the same time.
+        rngs = [_rand.rand64bits() for _ in range(200000)]
+        q.put(rngs)
+
+    ts = [threading.Thread(target=_target) for _ in range(5)]
+
+    for t in ts:
+        t.start()
+
+    for t in ts:
+        t.join()
+
+    ids = set()
+
+    while not q.empty():
+        new_ids_list = q.get()
+
+        new_ids = set(new_ids_list)
+        assert len(new_ids) == len(new_ids_list), "Collision found in ids"
+        assert ids & new_ids == set()
+        ids = ids | new_ids
+
+    assert len(ids) > 0
+
+
+def test_tracer_usage_fork():
+    q = mp.Queue()
+    pid = os.fork()
+
+    # Similar test to test_fork() above except we use the tracer API.
+    # In this case we expect to never have collisions.
+    if pid > 0:
+        # parent
+        parent_ids_list = list(
+            chain.from_iterable((s.span_id, s.trace_id) for s in [tracer.start_span("s") for _ in range(1000)])
```

Comment:
these are all going to be apart of the same trace right?

would it be better to create 1k individual traces with 1 span each? that way you can validate that no id generated our of 2k is the same, vs this which validates that no id inside of 1,001 ids are the same?

Reply:
same thought for below in `usage_multiprocess` as well.

Reply:
Good almost-catch! 🙂

`start_span()` doesn't pick up the active context (`tracer.trace()` does). This is confirmed in the tests since we check for duplicates across _all_ ids generated. So if the trace_ids were being duplicated, the check on line 148 would fail.